Add credits roll

KoBeWi opened the pull request:
Godot is said to be a GAME engine, but it always felt like there is not enough GAME in it. This PR fixes that by adding credits roll:

https://github.com/godotengine/godot/assets/2223172/ef55eb52-961e-43f0-9100-a724072f4093

@akien-mga I remembered ;)

KoBeWi committed:
Add credits roll

KoBeWi commented:
Now that April Fools is over, I restored the About dialog and the credits roll is instead activated when double clicking Remi:

https://github.com/godotengine/godot/assets/2223172/2036e761-3901-45f0-80ab-453c3f5cc828

The change is pretty self-contained, so it can be merged as easter egg. Especially when, you know, it's Easter time 😏

fire commented:
I though of a way to make this fit Godot Engine.

There is a slider toggle for auto scroll that is enabled by default. On about page load we scroll.

KoBeWi commented:
Nah, I can tweak some details like how it's activated, how fast it scrolls or its colors, but the screen itself is going to stay as is. If it can't be merged in the current form, feel free to close it.

KoBeWi commented:
Rebased for the first anniversary.
I simplified the code and handled language changes.

Unfortunately the credits popup doesn't work properly .-.
In single window mode it only takes half of the screen.
In multiwindow mode it... completely freezes my system for a few seconds, to the point that even sound stops playing. Before it freezes completely, the popup is moving to the left?? What's more, the glitch is caused by #88852

Calinou commented:
Tested locally on Linux/X11 (with single-window mode disabled), I can't get the credits to show up by double-clicking Rémi's name in the About dialog. Nothing happens when double-clicking. It works with single-window mode enabled though.

I think holding left-click in the dialog should speed up scrolling just like <kbd>Space</kbd> and <kbd>Enter</kbd> do, so you can do it with touch-based input. Right-clicking could exit the credits like pressing <kbd>Escape</kbd>.

Calinou commented:
> Is right click possible with touch input? Maybe I should just remove the credits from Android editor, too much hassle to test it. That would make mouse controls unnecessary I assume.

No, but as long as using the back button on the device exits the credits, it should be fine.

KoBeWi commented:
I added mouse controls and check for `NOTIFICATION_WM_GO_BACK_REQUEST` (not tested).

> A 64x64 window is created in the top-left corner of the screen.

That's because of the condition that checks for embedded window.
But this confirms that fullscreen is a problem for some reason. I will just make it windowed :/

Please test again.

Calinou reviewed (commented):
Works great on Linux/X11 now, both with single window mode disabled and enabled :slightly_smiling_face: 

One issue I noticed is that now that it's in a window, having the credits in a small window will result in just a black screen being visible for a while (until the Godot logo becomes visible). This is not an issue in a fullscreen or maximized editor, which hints that the screen size is being used to place the initial logo position instead of the window size.

Also, the logo is a bit blurry due to being blown to a high size. This is on a 4K display at 100% editor scale:

![image](https://github.com/user-attachments/assets/f1caec35-8c2a-443c-b18b-8a490010e05e)

I suppose this could be fixed by increasing the SVG scale of the source image, and downscaling it in the About dialog. Either way, it's not critical.

Calinou commented on editor/credits_roll.cpp:
```diff
@@ -0,0 +1,235 @@
+			label->add_theme_font_size_override(SNAME("font_size"), font_size_normal);
+			break;
+
+		case LabelSize::HEADER:
+			label->add_theme_font_size_override(SNAME("font_size"), font_size_header);
+			break;
+
+		case LabelSize::BIG_HEADER:
+			label->add_theme_font_size_override(SNAME("font_size"), font_size_big_header);
+			break;
```
Headers and big headers could use the bold font for better readability, like we do elsewhere in the editor.

KoBeWi commented:
Fixed everything except the logo. It's just an editor icon, not sure if it can be easily adjusted.

Calinou commented:
It works well now, but if you resize the editor window to a small size after the credits were shown once, it will no longer be center-aligned:

![image](https://github.com/user-attachments/assets/4a57cd44-6d8d-4f84-9598-4bc695a696b5)

This seems to occur with a 1366x768 editor window, for instance. 1920x1080 looks OK but even 1600x900 is slightly misaligned.

KoBeWi commented:
That's because some labels are wider than window size. Not easily fixable.

Repiteo reviewed (approved):
lmaooooooooo

AThousandShips commented on editor/credits_roll.cpp:
```diff
@@ -0,0 +1,245 @@
+
+		case LabelSize::HEADER: {
+			label->add_theme_font_size_override(SNAME("font_size"), font_size_header);
+			label->add_theme_font_override(SNAME("font"), bold_font);
```
```suggestion
			label->add_theme_font_override(SceneStringName(font), bold_font);
```
For all

KoBeWi commented:
I see it was marked as Feature for blog post.
Easter eggs are supposed to be hidden, so maybe we shouldn't mention it at all? 🤔

Repiteo commented:
Thanks!
